airbyte-scheduler/src/main/java/io/airbyte/scheduler/temporal/SpecWorkflow.java

```diff
@@ -91,46 +89,29 @@ public SpecActivityImpl(ProcessBuilderFactory pbf, Path workspaceRoot) {
       try {
-        // todo (cgardens) - we need to find a way to standardize log paths sanely across all workflow.
-        // right now we have this in temporal workflow.
+        // todo (cgardens) - there are 2 sources of truth for job path. we need to reduce this down to one,
+        // once we are fully on temporal.
         final Path jobRoot = workspaceRoot
-            .resolve("spec")
-            .resolve(launcherConfig.getDockerImage().replaceAll("[^A-Za-z0-9]", ""))
-            .resolve(String.valueOf(Instant.now().getEpochSecond()));
+            .resolve(String.valueOf(launcherConfig.getJobId()))
+            .resolve(String.valueOf(launcherConfig.getAttemptId().intValue()));
+
+        MDC.put("job_id", String.valueOf(launcherConfig.getJobId()));
+        MDC.put("job_root", jobRoot.toString());
+        MDC.put("job_log_filename", WorkerConstants.LOG_FILENAME);
 
         final IntegrationLauncher integrationLauncher =
             new AirbyteIntegrationLauncher(launcherConfig.getJobId(), launcherConfig.getAttemptId().intValue(), launcherConfig.getDockerImage(), pbf);
-        final Process process = integrationLauncher.spec(jobRoot).start();
-
-        LineGobbler.gobble(process.getErrorStream(), LOGGER::error);
-
-        final AirbyteStreamFactory streamFactory = new DefaultAirbyteStreamFactory();
-
-        Optional<ConnectorSpecification> spec;
-        try (InputStream stdout = process.getInputStream()) {
-          spec = streamFactory.create(IOs.newBufferedReader(stdout))
-              .filter(message -> message.getType() == AirbyteMessage.Type.SPEC)
-              .map(AirbyteMessage::getSpec)
-              .findFirst();
-
-          // todo (cgardens) - let's pre-fetch the images outside of the worker so we don't need account for
-          // this.
-          // retrieving spec should generally be instantaneous, but since docker images might not be pulled
-          // it could take a while longer depending on internet conditions as well.
-          WorkerUtils.gentleClose(process, 30, TimeUnit.MINUTES);
-        }
 
-        int exitCode = process.exitValue();
-        if (exitCode == 0) {
-          if (spec.isEmpty()) {
-            throw new RuntimeException("Spec job failed to output a spec struct.");
-          } else {
-            return spec.get();
-          }
+        final JobGetSpecConfig jobGetSpecConfig = new JobGetSpecConfig().withDockerImage(launcherConfig.getDockerImage());
+        final OutputAndStatus<JobOutput> run = new JobOutputGetSpecWorker(new DefaultGetSpecWorker(integrationLauncher))
+            .run(jobGetSpecConfig, jobRoot);
+        if (run.getStatus() == JobStatus.SUCCEEDED) {
+          Preconditions.checkState(run.getOutput().isPresent());
+          return run.getOutput().get();
         } else {
-          throw new RuntimeException(String.format("Spec job subprocess finished with exit code %s", exitCode));
+          throw new TemporalJobException();
```

Comment:
Same

Reply:
Can you add some tests for these classes? That would likely have caught this bug